File: dp/c/policy_based_design.c
// Policy-Based Design in C
// Compose behavior via interchangeable function pointer tables

#include <stdio.h>
#include <string.h>
#include <stdbool.h>

// Log policy
typedef struct { void (*log)(const char* msg); } LogPolicy;

void console_log(const char* msg) { printf("[Console] %s\n", msg); }
void silent_log(const char* msg) { (void)msg; }

static const LogPolicy CONSOLE_LOG = {console_log};
static const LogPolicy SILENT_LOG = {silent_log};

// Validation policy
typedef struct { bool (*validate)(const char* value, char* err, int err_size); } ValidationPolicy;

bool strict_validate(const char* value, char* err, int err_size) {
    if (!value || value[0] == '\0') { snprintf(err, err_size, "empty"); return false; }
    if (strlen(value) > 100) { snprintf(err, err_size, "too long"); return false; }
    if (strchr(value, '<')) { snprintf(err, err_size, "HTML not allowed"); return false; }
    return true;
}

bool lenient_validate(const char* value, char* err, int err_size) {
    if (!value || value[0] == '\0') { snprintf(err, err_size, "empty"); return false; }
    return true;
}

static const ValidationPolicy STRICT = {strict_validate};
static const ValidationPolicy LENIENT = {lenient_validate};

// Composed service
typedef struct {
    const LogPolicy* logger;
    const ValidationPolicy* validator;
} ConfigService;

bool config_set(ConfigService* svc, const char* key, const char* value) {
    char msg[128], err[64];
    snprintf(msg, sizeof(msg), "Setting %s = %s", key, value);
    svc->logger->log(msg);

    if (!svc->validator->validate(value, err, sizeof(err))) {
        snprintf(msg, sizeof(msg), "Validation failed: %s", err);
        svc->logger->log(msg);
        return false;
    }

    snprintf(msg, sizeof(msg), "Saved %s", key);
    svc->logger->log(msg);
    return true;
}

int main(void) {
    printf("=== Strict + Console ===\n");
    ConfigService strict_svc = {&CONSOLE_LOG, &STRICT};
    config_set(&strict_svc, "name", "Alice");
    config_set(&strict_svc, "bad", "<script>");

    printf("\n=== Lenient + Silent ===\n");
    ConfigService lenient_svc = {&SILENT_LOG, &LENIENT};
    bool ok = config_set(&lenient_svc, "html", "<b>bold</b>");
    printf("Result: %s\n", ok ? "saved" : "failed");
    return 0;
}

File: dp/rust/src/policy_based_design.rs
// Policy-Based Design in Rust
// Compose behavior from interchangeable policy traits via generics

use std::fmt;

// --- Logging policy ---
trait LogPolicy {
    fn log(&self, msg: &str);
}

struct ConsoleLog;
impl LogPolicy for ConsoleLog {
    fn log(&self, msg: &str) { println!("[Console] {}", msg); }
}

struct SilentLog;
impl LogPolicy for SilentLog {
    fn log(&self, _msg: &str) {} // no-op
}

// --- Validation policy ---
trait ValidationPolicy {
    fn validate(&self, value: &str) -> Result<(), String>;
}

struct StrictValidation;
impl ValidationPolicy for StrictValidation {
    fn validate(&self, value: &str) -> Result<(), String> {
        if value.is_empty() { return Err("empty value".into()); }
        if value.len() > 100 { return Err("too long".into()); }
        if value.contains('<') { return Err("HTML not allowed".into()); }
        Ok(())
    }
}

struct LenientValidation;
impl ValidationPolicy for LenientValidation {
    fn validate(&self, value: &str) -> Result<(), String> {
        if value.is_empty() { return Err("empty value".into()); }
        Ok(())
    }
}

// --- Storage policy ---
trait StoragePolicy {
    fn save(&mut self, key: &str, value: &str);
    fn load(&self, key: &str) -> Option<String>;
}

struct InMemoryStorage {
    data: std::collections::HashMap<String, String>,
}

impl InMemoryStorage {
    fn new() -> Self { InMemoryStorage { data: std::collections::HashMap::new() } }
}

impl StoragePolicy for InMemoryStorage {
    fn save(&mut self, key: &str, value: &str) {
        self.data.insert(key.to_string(), value.to_string());
    }
    fn load(&self, key: &str) -> Option<String> {
        self.data.get(key).cloned()
    }
}

// --- Composed service using policies ---
struct ConfigService<L: LogPolicy, V: ValidationPolicy, S: StoragePolicy> {
    logger: L,
    validator: V,
    storage: S,
}

impl<L: LogPolicy, V: ValidationPolicy, S: StoragePolicy> ConfigService<L, V, S> {
    fn new(logger: L, validator: V, storage: S) -> Self {
        ConfigService { logger, validator, storage }
    }

    fn set(&mut self, key: &str, value: &str) -> Result<(), String> {
        self.logger.log(&format!("Setting {} = {}", key, value));
        self.validator.validate(value)?;
        self.storage.save(key, value);
        self.logger.log(&format!("Saved {}", key));
        Ok(())
    }

    fn get(&self, key: &str) -> Option<String> {
        self.logger.log(&format!("Getting {}", key));
        self.storage.load(key)
    }
}

fn main() {
    // Strict + verbose config
    println!("=== Strict + ConsoleLog ===");
    let mut strict_svc = ConfigService::new(
        ConsoleLog, StrictValidation, InMemoryStorage::new(),
    );
    let _ = strict_svc.set("name", "Alice");
    let _ = strict_svc.set("bad", "<script>");
    println!("name = {:?}", strict_svc.get("name"));

    // Lenient + silent config
    println!("\n=== Lenient + SilentLog ===");
    let mut lenient_svc = ConfigService::new(
        SilentLog, LenientValidation, InMemoryStorage::new(),
    );
    let _ = lenient_svc.set("html", "<b>bold</b>");
    println!("html = {:?}", lenient_svc.get("html"));
}
